enhancements/20230907-snapshot-space-management.md

```diff
@@ -0,0 +1,94 @@
+# Snapshot Space Management
+
+## Summary
+
+This proposal leverages [LH Snapshot Space Management](https://github.com/longhorn/longhorn/issues/6563) to control snapshot space usage in the system. LH introduces two new attributes `snapshotMaxCount` and `snapshotMaxSize` in Volume. Harvester can use these parameters to set upper bound for snapshot space usage.
+
+### Related Issues
+
+https://github.com/harvester/harvester/issues/4478
+
+## Motivation
+
+### Goals
+
+- Set snapshot count and size limitations for each volume.
+
+## Proposal
+
+Use a PVC controller to watch PVCs. If there is `harvesterhci.io/snapshotMaxCount` or `harvesterhci.io/snapshotMaxSize` annotation on a PVC, the system uses the value to update the related LH Volume.
+
+### User Stories
+
+With snapshot space management, snapshot space usage is controllable. Users can set snapshot limitation and evaluate maximum space usage.
+
+### User Experience In Detail
+
+Before snapshot space management:
+The default maximum snapshot count for each volume is `250`. This is a constants value in the [source code](https://github.com/longhorn/longhorn-engine/blob/8b4c80ab174b4f454a992ff998b6cb1041faf63d/pkg/replica/replica.go#L33) and users don't have a way to control it. If a volume size is 1G, the maximum snapshot space usage will be 250G.
+
+After snapshot space management:
+There is a configurable default maximum snapshot count setting. Users can update it to overwrite the fixed value in the system.
+Users can set different maximum snapshot count and size on each volume. A more important volume can have more snapshot space usage.
```

Comment:
With regards to user experience, could we also plan for the UI to warn the user if something like the following scenario occurs:

* Volume has 10 successful snapshots
* User changes `snapshot-max-count` to a value <= 10

The UI could display a warning banner stating that the new value prevents snapshots from being executed on volumes whose snapshot amount >= the new value?

Reply:
Yes, it's good suggestion to notify users when there is not enough count or size. At first version, I would like to forbid users from narrowing down `snapshot-max-count` & `snapshot-max-size`. In LH, system also can create snapshot for rebuilding a volume. If we allow to narrow down count or size, the system may become unstable.

Reply:
After some implementations, we have to allow users to narrow down `snapshot-max-count` & `snapshot-max-size` in LH. The reason is that we don't operate LH Volume CR directly. We use PVC and StorageClass for it. If we use annotations on PVC to control `snapshot-max-count` & `snapshot-max-size` in LH Volume, we have to allow narrowing down operations.

In LH, it will use `Engine.Status.Snapshots` information to check whether new `snapshot-max-count` & `snapshot-max-size` are valid. However, the information from `Engine.Status.Snapshots` may be stale, because it's updated [every 5 seconds](https://github.com/longhorn/longhorn-manager/blob/e9f5eac31f56056649bc45a2d740a9d6d10cadd8/controller/engine_controller.go#L51). In some cases, the real snapshot usage may be more than `snapshot-max-count` & `snapshot-max-size`. For example, a user create a new snapshot and another user narrows down `snapshot-max-count`. At this moment, the new snapshot hasn't been reflected on Engine CR, so the system will allow the operation. In this case, users don't break the system, but the real snapshots usage is more than `snapshot-max-count` & `snapshot-max-size`.